mm2src/coins/utxo/utxo_common.rs

```diff
@@ -1896,26 +1949,77 @@ pub fn watcher_validate_taker_payment<T: UtxoCommonOps + SwapOps>(
+    let taker_payment_refund_preimage: UtxoTx = try_f!(deserialize(input.taker_payment_refund_preimage.as_slice()));
+    let taker_pub = &try_f!(
         Public::from_slice(&input.taker_pub).map_err(|err| ValidatePaymentError::InvalidParameter(err.to_string()))
     );
-    let second_pub = &try_f!(
+    let maker_pub = &try_f!(
         Public::from_slice(&input.maker_pub).map_err(|err| ValidatePaymentError::InvalidParameter(err.to_string()))
     );
-    validate_payment(
-        coin.clone(),
-        tx,
-        DEFAULT_SWAP_VOUT,
-        first_pub,
-        second_pub,
-        &input.secret_hash,
-        input.amount,
-        input.time_lock,
-        input.try_spv_proof_until,
-        input.confirmations,
-    )
+    let expected_redeem = payment_script(input.time_lock, &input.secret_hash, taker_pub, maker_pub);
+    let coin = coin.clone();
+
+    let fut = async move {
+        let inputs_signed_by_pub = check_all_utxo_inputs_signed_by_pub(&taker_payment_tx, &input.taker_pub)
+            .map_to_mm(ValidatePaymentError::InternalError)?;
+        if !inputs_signed_by_pub {
+            return MmError::err(ValidatePaymentError::WrongPaymentTx(
+                "Taker payment does not belong to the verified public key".to_string(),
+            ));
+        }
+
+        let taker_payment_locking_script = match taker_payment_tx.outputs.get(DEFAULT_SWAP_VOUT) {
+            Some(output) => output.script_pubkey.clone(),
+            None => {
+                return MmError::err(ValidatePaymentError::WrongPaymentTx(
+                    "Payment tx has no outputs".to_string(),
+                ))
+            },
+        };
+
+        if taker_payment_locking_script != Builder::build_p2sh(&dhash160(&expected_redeem).into()).to_bytes() {
+            return MmError::err(ValidatePaymentError::WrongPaymentTx(format!(
+                "Payment tx locking script {:?} doesn't match expected",
+                taker_payment_locking_script
+            )));
+        }
+
+        let script_sig = match taker_payment_refund_preimage.inputs.get(DEFAULT_SWAP_VIN) {
+            Some(input) => Script::from(input.script_sig.clone()),
+            None => {
+                return MmError::err(ValidatePaymentError::WrongPaymentTx(
+                    "Taker payment refund tx has no inputs".to_string(),
+                ))
+            },
+        };
+
+        let instruction = script_sig
+            .iter()
+            .last()
+            .or_mm_err(|| ValidatePaymentError::WrongPaymentTx(String::from("Instruction not found")))?
+            .map_to_mm(|err| ValidatePaymentError::WrongPaymentTx(err.to_string()))?;
+
+        let redeem_script = instruction.data.or_mm_err(|| {
+            ValidatePaymentError::WrongPaymentTx(String::from("No redeem script in the taker payment refund preimage"))
+        })?;
+
+        if expected_redeem.to_bytes().take() != redeem_script {
```

Comment:
`to_bytes` clones the internal data, which isn't required here to compare with `redeem_script` slice. Could you please add `as_slice` method to `Script` instead?